blast/ncbi/doc/man/getfeat.1
.TH GETFEAT 1 2002-08-23 NCBI "NCBI Tools User's Manual"
.SH NAME
getfeat \- extract information about sequences' coding regions
.SH SYNOPSIS
.B getfeat
[\|\fB\-\fP\|]
[\|\fB\-a\fP\|]
[\|\fB\-b\fP\|]
[\|\fB\-g\fP\ \fIN\fP\|]
[\|\fB\-i\fP\ \fIstr\fP\|]
[\|\fB\-l\ F\fP\|]
[\|\fB\-o\fP\ \fIfilename\fP\|]
[\|\fB\-p\fP\|]
[\|\fB\-r\fP\ \fIN\fP\|]
[\|\fB\-s\ F\fP\|]
.SH DESCRIPTION
\fBgetfeat\fP retrieves biological sequences over the network and
extracts information about their coding regions.
.SH OPTIONS
A summary of options is included below.
.TP
\fB\-\fP
Print usage message
.TP
\fB\-a\fP
ID is GenBank accession
.TP
\fB\-b\fP
ID is PIR accession
.TP
\fB\-g\fP\ \fIN\fP
GenInfo Import ID
.TP
\fB\-i\fP\ \fIstr\fP
Input string id (default = \fBANANIFDR1\fP)
.TP
\fB\-l\ F\fP
ID is NOT GenBank locus
.TP
\fB\-o\fP\ \fIfilename\fP
Output File Name (default = \fBgetfeat.out\fP)
.TP
\fB\-p\fP
ID is PIR locus
.TP
\fB\-r\fP\ \fIN\fP
Return type:
.RS
.PD 0
.IP 1
Bioseq
.IP 2
Seg-set
.IP 3
Nuc-prot (default)
.IP 4
Pub-set
.PD
.RE
.TP
\fB\-s\ F\fP
Do not output a Seq-entry.
.SH AUTHOR
The National Center for Biotechnology Information.


QuickTimeStreamingServer/ReleaseNotes.txt

                                 RELEASE NOTES
                             Darwin Streaming Server 

These release notes are for the current development version of Darwin
Streaming Server.

------------------------------------------------------------------------
Date: 8/7/2001
Version: 3.0.1+development [v349]
cvs tag: qtssServer-349

*   Updated build platforms: Solaris 8, Redhat Linux 7.1, FreeBSD 4.3, 
    Mac OS X 10.1
*   Retransmit/overbuffering (Skip Protection) algorithms tuned
    substantially
*   "reliable_udp" attribute defaults to "enabled" 
*   Various WebAdmin fixes for Windows platform

------------------------------------------------------------------------
Date: 8/9/2001
Version: 3.0.1+development [v350]
cvs tag: qtssServer-350

*   Minor bug fixes.

------------------------------------------------------------------------
Date: 8/14/2001
Version: 3.0.1+development [v351]
cvs tag: qtssServer-351

*   QTSSFileModule - enabled caching based on preferences (shared and
    private).
*   Add buffer delay to on demand movies.
*   QTFile - code cleanup.
*   Changed some paths for the installer.
*   Added support for MP3Broadcaster on Windows.

------------------------------------------------------------------------
Date: 8/16/2001
Version: 3.0.1+development [v352]
cvs tag: qtssServer-352

*   Added MP3StreamingModule to all builds.
*   Fixed gdb script to execute QTSS from build dir on MacOSX.

------------------------------------------------------------------------
Date: 8/19/2001
Version: 3.0.1+development [v353]
cvs tag: qtssServer-353

*   Added:
        add_seconds_to_client_buffer_delay -- Increases client buffer
                                delay before playing movie.
        record_movie_file_sdp -- Records a .sdp file for a .mov file.
                                Appears as moviename.mov.sdp
        enable_movie_file_sdp -- Plays a .mov.sdp file for a .mov movie.
*   Fixed server to play .mov.sdp files.
*   QTSSFileModule enables both private and shared movie cache buffers.

------------------------------------------------------------------------
Date: 8/23/2001
Version: 3.0.1+development [v354]
cvs tag: qtssServer-354

*   Added prefs for MP3StreamingModule.
*   Default Reliable UDP dir is / on UNIX and \ on NT
*   Use Enable Reliable UDP pref to disable Reliable UDP
*   Included new Web Admin
*   Changed relay preferences to XML format.
        Old relay config files will not work.

------------------------------------------------------------------------
Date: 08/28/2001
Version: 3.0.1+development [v355]
cvs tag: qtssServer-355

*	Added preference:
	 RTSP_error_message to enable debugging error strings returned in 
	 the content body of RTSP error responses.
	
*	Changed RTSP code to:
	1) require a Cseq field for OPTIONS requests
	2) allow missing Range headers in a PLAY request (defaults to entire movie)
	3) add content length header to Error responses containing error messages
	4) add SDP o= and t= fields to on-demand movie Describe responses

*      Changed Relay to relay incoming ANNOUNCEd broadcasts
       1) All ANNOUNCEd broadcasts, or
       2) All broadcasts from an IP address
       3) A broadcast URL from an IP address

       (a) Can still configure the relay to relay UDP streams
       (b) Can pull an SDP or movie from source and relay it

*      Changed Relay to ANNOUNCE the SDPs to its destinations
       1) for ANNOUNCED sources
       2) for sources that are pulled from the source server

*      Changed Relay config file to be in XML format
       1) relayconfig.xml-Sample will be installed instead of old streamingrelay.conf

*	Added General Settings and Change Password to UI
	1) Require existing password to change to new
	2) Require the user to enter the same password twice
	3) Change the auth cookie so the user won't have to re-authenticate

-------------------------------------------------------------------------------------------------
Date: 8/31/2001
Version: 3.0.1+development [v356]
cvs tag: qtssServer-356

*	 Admin Module change
	 1) Added an admin module pref called "AdministratorGroup" with
	 a default value of "admin"
	 2) Changed Admin Module to allow all users that belong to the
	 AdministratorGroup, instead of allowing only one user called
	 "streamingadmin".
	 3) A default qtgroups gets installed from now. The default qtusers
	 file has user 'streamingadmin' with password 'default'. The default
	 qtgroups file has the group 'admin' with the user 'streamingadmin'.
	 4) The login name for the web admin can be changed by adding the user
	 to the qtusers file and adding the user to the 'admin' group in the
	 qtgroups file.

*	Added "RTSP_debug_printfs" server pref to show the incoming and outgoing RTSP.
	Fixed Session header response to remove ";timeout=" field if the value is not defined.
	Fixed .mov.sdp Describe response to include the Last-Modified and Cache-Control headers.
	Fixed RTSP Session header parser to also stop on a ';'
	Changes shared file buffer size to default to maximum size if the bit-rate is 0 and no fixed 
	buffer size is defined.

-------------------------------------------------------------------------------------------------
Date: 9/04/2001
Version: 3.0.1+development [v357]
cvs tag: qtssServer-357

*	more implementation code for Web admin added.
*	New XML parser for reading prefs.
*	full implementation of MP3 streaming module is included.

-------------------------------------------------------------------------------------------------
Date: 9/06/2001
Version: 3.0.1+development [v359]
cvs tag: qtssServer-359

*	Relay that pulls source from the same server can be setup now
	[rtsp_source SOURCE with own IP address as the source_addr]
   
-------------------------------------------------------------------------------------------------
Date: 9/10/2001
Version: 3.0.1+development [v361]
cvs tag: qtssServer-361

*	UI: play and stop buttons in playlist screen now start/stop playlists immediately

blast/ncbi/demo/getfeat.c
/*****************************************************************************
*
*   getfeat.c
*   	entrez version
*
*****************************************************************************/
#include <accentr.h>
#include <seqport.h>

#define NUMARGS 9
Args myargs[NUMARGS] = {
 /*0*/ { "Input string id", "ANANIFDR1", NULL, NULL, FALSE, 'i', ARG_STRING, 0.0,0,NULL},
 /*1*/ { "Output Seq-entry", "T", NULL, NULL, TRUE , 's', ARG_BOOLEAN, 0.0,0,NULL},
 /*2*/ { "Output File Name", "getfeat.out", NULL, NULL, FALSE, 'o', ARG_FILE_OUT, 0.0,0,NULL},
 /*3*/ { "GenInfo Import id", "0", NULL, NULL, TRUE , 'g', ARG_INT, 0.0,0,NULL},
 /*4*/ { "Id is GenBank accession", "F", NULL, NULL, TRUE , 'a', ARG_BOOLEAN, 0.0,0,NULL},
 /*5*/ { "Id is GenBank locus", "T", NULL, NULL, TRUE , 'l', ARG_BOOLEAN, 0.0,0,NULL},
 /*6*/ { "Id is PIR accession", "F", NULL, NULL, TRUE , 'b', ARG_BOOLEAN, 0.0,0,NULL},
 /*7*/ { "Id is PIR locus", "F", NULL, NULL, TRUE , 'p', ARG_BOOLEAN, 0.0,0,NULL},
 /*8*/ { "Return type 1=Bioseq,2=seg-set,3=nuc-prot,4=pub-set", "3", "1", "4", TRUE , 'r', ARG_INT, 0.0,0,NULL}};

/*
** Prototype for Coding Region exploring function
*/

static void LIBCALLBACK GetSeqFeat(AsnExpOptStructPtr aeosp);

/*
** We will need to pass both a file pointer and an ASN I/O stream pointer
** to the exploring function but we can only pass one pointer. We thus 
** combine them into a single structure and pass that pointer instead.
**
** Our exploring function, "GetSeqFeat", will now have access the both
** elements of this structure.
*/

typedef struct demostruct {
	FILE * fp;
	AsnIoPtr aip;
} DemoStruct;

/*
** Our Feature Presentation...
*/

Int2 Main(void)
{
	Boolean         seqentry; /* Arg 1 should always be TRUE */
	Int2            retcode;  /* Default is nuc-prot        */

	DemoStruct PNTR dsp;      /* Struct ptr used to pass data */

	TextSeqIdPtr    tsip;     /* Text object, name, locus, or release */
	SeqIdPtr        sip;      /* Same as ValNodePtr, generic data ptr */

	GiimPtr         giim;     /* GenInfo struct passed via "sip" */

	FILE* fp;
	AsnExpOptPtr aeop;
	AsnIoPtr aip;

	Int4 uid;
	SeqEntryPtr sep;
	Boolean is_network;

	/*
	** Get program arguments
	*/

	if ( !GetArgs("GetSeq 1.0", NUMARGS, myargs) ) return 1;

	/*
	** We'll be working with SeqEntry objects, so load the
	** the loaders here, this also gets the sequence alphabets.
	*/

	if ( !SeqEntryLoad() ) return 1;

	/*
	** Set parameters from the command line
	*/

	seqentry = (Boolean)myargs[1].intvalue;
	retcode = (Int2)myargs[8].intvalue;

	/*
	** Create a DemoStruct object, we will use this struct
	** to pass two other data elements to the exploring
	** function.
	*/

	dsp = MemNew(sizeof(DemoStruct));

	/*
	** Instantiate ASN stream pointers and data holders.
	** "sip" is used to pass the information in "tsip".
	*/

	tsip = TextSeqIdNew();
	sip = ValNodeNew(NULL);
	sip->data.ptrvalue = tsip;

	/*
	** From the command line, set either the name or accession field
	** and the choice field of the SeqIdPtr, this structure will be
	** used to return an ASN object from the CD-ROM or Network.
	*/

	if (myargs[3].intvalue)  {                /* giim */
		giim = GiimNew();
		giim->id = myargs[3].intvalue;
		sip->data.ptrvalue = giim;
		sip->choice = SEQID_GIIM;
	} else if (myargs[4].intvalue) {          /* GenBank accession */
		tsip->accession = myargs[0].strvalue;
		sip->choice = SEQID_GENBANK;
	} else if (myargs[5].intvalue) {          /* GenBank locus name */
		tsip->name = myargs[0].strvalue;
		sip->choice = SEQID_GENBANK;
	} else if (myargs[6].intvalue) {          /* PIR accession */
		tsip->accession = myargs[0].strvalue;
		sip->choice = SEQID_PIR;
	} else if (myargs[7].intvalue) {          /* PIR locus name */
		tsip->name = myargs[0].strvalue;
		sip->choice = SEQID_PIR;
	}

	/*	
	** Initialize, open and otherwise prepare for CD-ROM or Network access.
	*/

	if ( !EntrezInit("GetFeat", FALSE, &is_network) ) {
		ErrShow();
	}

	/*
	** Load the DemoStruct pointer
	*/

	fp = FileOpen(myargs[2].strvalue, "w");
	dsp->fp = fp;
	dsp->aip = AsnIoNew(ASNIO_TEXT_OUT, fp, NULL, NULL, NULL);

	/*
	** Note that no exploring is done at this point.  We are only 
	** setting up for the exploration.
	**
	** The exploration function, "GetSeqFeat", will be called each time
	** a "Seq-Feat" is encountered in the ASN data stream for this entry.
	*/

	aip = AsnIoNullOpen();
	if ((aeop = AsnExpOptNew(aip, "Seq-feat", (Pointer)dsp, GetSeqFeat)) == NULL)
		ErrPost(100, 1, "no aeop 1");

	if (seqentry) {
		uid = EntrezFindSeqId(sip);
		if ( !uid )
			ErrShow();
		else {
			sep = EntrezSeqEntryGet(uid, retcode);
			if (sep == NULL)
				ErrShow();
			else {
				SeqEntryAsnWrite(sep, aip, NULL);
				fflush(fp);
				AsnIoReset(aip);
				SeqEntryFree(sep);
			}
		}
	} else {
		Message(MSG_ERROR, "Oops");
	}

	AsnIoClose(aip);
	FileClose(fp);
	EntrezFini();
	return 0;
}


/***************************************************************************8
**
** The exploration function.  Used to find coding regions
*/

void LIBCALLBACK GetSeqFeat(AsnExpOptStructPtr aeosp)
{
	SeqFeatPtr sfp;
	SeqPortPtr spp;
	Uint1      residue;
	FILE*      fp;
	Int2       i;
	Char       buffer[101];
	DemoStruct PNTR dsp;

	/*
	** Make sure we are at the beginning of a sequence feature.
	** If so, then set "sfp" to this address.
	*/

	if (aeosp->dvp->intvalue != START_STRUCT) return;

	sfp = (SeqFeatPtr)aeosp->the_struct;

	/*
	** We are looking for coding regions, 3 is a cdregion. (Defined 
	** elsewhere). Anything else simply return.
	*/

	if (sfp->data.choice != 3) return;

	/*
	** The SeqLoc may be quite complex, whereas a SeqPort is a linear
	** region.  We will use single letter IUPAC nucleic acid symbols.
	**
	** This single call represents a very complex process. Thank you NCBI!
	**
	** Seq_code_iupacna is a symbol defined elsewhere.
	*/

	spp = SeqPortNewByLoc(sfp->location, Seq_code_iupacna);
	if (spp == NULL) return;

	/*
	** Get the output file pointer from the passed data
	** structure.
	*/

	dsp = (DemoStruct PNTR) aeosp->data;
	fp = dsp->fp;

	/*
	** Write "New CdRegion" to file, then write the annotation for
	** this region in ASN Print-Value format.
	**
	** Use the demo program, "getseq" to see the entire entry. For example,
	** the default for this program uses the genbank locus "ANANIFDR1".
	**
	**       getseq -i ananifdr1 -o anaifdr1.asn -l
	*/

	fprintf(fp, "\nNew CdRegion\n");
	SeqFeatAsnWrite(sfp, dsp->aip, NULL);
	AsnIoReset(dsp->aip);

	/*
	** Write out the coding region 60 characters per line
	*/

	i = 0;
 	while ((residue = SeqPortGetResidue(spp)) != SEQPORT_EOF) {
		if (! IS_residue(residue))
		{
			buffer[i] = '\0';
			fprintf(fp, "%s\n", buffer);
			i = 0;
			switch (residue)
			{
				case SEQPORT_VIRT:
					fprintf(fp, "[Gap]\n");
					break;
				case SEQPORT_EOS:
					fprintf(fp, "[EOS]\n");
					break;
				default:
					fprintf(fp, "[Invalid Residue]\n");
					break;
			}
		} else {
			buffer[i] = residue;
			i++;
			if (i == 60) {
				buffer[i] = '\0';
				fprintf(fp, "%s\n", buffer);
				i = 0;
			}
		}
	}

	/*
	** Write out any remaining buffer here.
	*/

	if (i) {
		buffer[i] = '\0';
		fprintf(fp, "%s\n", buffer);
	}


	/*
	** Write out EOF to mark the end of the coding region
	** Free up memory used by the ssp. Return.
	*/

	fprintf(fp, "EOF\n");
	SeqPortFree(spp);

	return;
}

